### PR TITLE
Added module for ztest and testing as well

### PR DESCRIPTION
**Pull request by cassielayden**
#### Reference issue

Progress made on feature request #13662 

#### What does this implement/fix?

Implementation adds a module for performing the z tests and a test suite for the Z-test module as well. 
Z-Test module is at scipy/stats/ztest.py
Test module is at scipy/stats/tests/test_ztest.py

#### Additional information

The implementation is not complete. The functions in the added module are correct and working when provided the input outlines. Comments are included throughout the code that explain how and where input should be taken in a user interface. Feel free to contact me with more questions about this. 
Additionally, I will attempt to implement the interface as I see possible, however I do believe another developer may be better fit for this task.

### CODE REVIEW

**Comment by tupui** [comment]
Thanks @cassielayden for submitting a PR. I am afraid I will have to close it for a few reasons.

From the issue this PR is addressing, @charlotte12l already expressed some interest in working on this. While there is no formal assignment of issues/PR, we are all volunteers and try not to re-do the same work. We certainly don't want to have contributors competing on PRs. Instead we prefer to share the work load by discussing in issues or on the mail list before.

Thus, I would invite you to contact @charlotte12l and see how you can be of help if this topic interests you. In the present case, a new statistical test, we also would want to design if in a way we could reuse existing components from other existing tests. In terms of maintainability, we would not want to include an implementation which would be totally contained here.

Before working on something else, I would also invite you to read our [contributing documentation](https://scipy.github.io/devdocs/dev/contributor/contributor_toc.html). It's a long document, but it's really important that you understand it and follow it. There are valuable informations such as how to format your code and tests. In the end we will have to respect this document from start to finish before considering including any contribution. As for other issues you could have a look at, we do have a label [good first issue](https://github.com/scipy/scipy/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22).

Thank you for your understanding and your willingness to contribute to SciPy. While this PR was not successful, I do hope that you would consider contributing again.

**Comment by ilayn** [comment]
It's better to first check if @charlotte12l is actually working on it. Looks like it is stalled over there so it doesn't have to be closed in case they don't ping back.

**Comment by rgommers** [comment]
Thanks for reopening @ilayn. @tupui this PR needs some work but it's totally valid to work on this. I suggest only closing a PR if you're really certain that's the right thing to do (like, it stalled for a year and someone else took it over in a new PR, or the bug it addressed was already fixed.

File: scipy/stats/ztest.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+# This software is provided "as-is".  There are no expressed or implied
+# warranties of any kind, including, but not limited to, the warranties
+# of merchantability and fitness for a given application.  In no event
+# shall Gary Strangman be liable for any direct, indirect, incidental,
+# special, exemplary or consequential damages (including, but not limited
+# to, loss of use, data or profits, or business interruption) however
+# caused and on any theory of liability, whether in contract, strict
+# liability or tort (including negligence or otherwise) arising in any way
+# out of the use of this software, even if advised of the possibility of
+# such damage.
```
Comment (rgommers) [inline scipy/stats/ztest.py:17]: This is a new file, so it does not need to have this copyright copied over from another file. Please remove this block comment.

File: scipy/stats/ztest.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                              rng_integers, float_factorial)
+import scipy.special as special
+from scipy import linalg
+from . import distributions
+from . import mstats_basic
+from ._stats_mstats_common import (_find_repeats, linregress, theilslopes,
+                                   siegelslopes)
+from ._stats import (_kendall_dis, _toint64, _weightedrankedtau,
+                     _local_correlations)
+from dataclasses import make_dataclass
```
Comment (rgommers) [inline scipy/stats/ztest.py:50]: Please import only the functions and modules you are actually using to implement the Z test functionality. Running `flake8` on this file will show you unnecessary imports.

File: scipy/stats/ztest.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+           'SpearmanRConstantInputWarning', 'spearmanr', 'pointbiserialr',
+           'kendalltau', 'weightedtau', 'multiscale_graphcorr',
+           'linregress', 'siegelslopes', 'theilslopes', 'ttest_1samp',
+           'ttest_ind', 'ttest_ind_from_stats', 'ttest_rel',
+           'kstest', 'ks_1samp', 'ks_2samp',
+           'chisquare', 'power_divergence', 'mannwhitneyu',
+           'tiecorrect', 'ranksums', 'kruskal', 'friedmanchisquare',
+           'rankdata',
+           'combine_pvalues', 'wasserstein_distance', 'energy_distance',
+           'brunnermunzel', 'alexandergovern']
```
Comment (rgommers) [inline scipy/stats/ztest.py:76]: Same here, just add `__all__ = ['ztest']` (if that's the function name).

File: scipy/stats/ztest.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+_one_sample_z_table = {
```
Comment (rgommers) [inline scipy/stats/ztest.py:77]: Can you add a comment about where these values come from?

Also, is 4 digits precise enough, and is there not a way to calculate these values instead?

**Review by rgommers — CHANGES_REQUESTED** [review]
A few high-level comments to get started.

I suggest first writing the public function, something like
```
def ztest(...):
    """
    Docstring here
    """
    ...
```

**Comment by charlotte12l** [comment]
@tupui @ilayn @rgommers Sorry for the late reply. I'm currently not working on it because there is not much interest after I mailed the mailinglist about this, so I added this to the end of my GSoC timeline.
@cassielayden I read your codes and I'd recommend taking a look at [gh-12873](https://github.com/scipy/scipy/pull/12873) for a good example of implementing a new hypothesis test(Thanks @mdhaber for mentioning the example in https://github.com/scipy/scipy/issues/13662 ). 
Moreover, maybe it is better to return the p-values rather than 0/1 to make it consistent with other tests? You can use `_normtest_finish` to get the p-value from the Z statistic. 
Finally, z-test is quite similar to t-test, I think you can always refer to t-test implementation for naming, code structures, tests, logics, etc. :)

**Comment by tupui** [comment]
> @tupui @ilayn @rgommers Sorry for the late reply. I'm currently not working on it because there is not much interest after I mailed the mailinglist about this, so I added this to the end of my GSoC timeline.

No worries, and thanks for helping with these specific information 😃

File: scipy/stats/ztest.py (hunk 4 above)
Comment (bashtage) [reply scipy/stats/ztest.py:77]: These are the cdf of a normal.  Why not use `stats.norm.cdf`?

**Pull request closed without merging**